controllers/k8ssandra/k8ssandracluster_controller.go

```diff
@@ -602,21 +324,358 @@ func (r *K8ssandraClusterReconciler) setStatusForStargate(kc *api.K8ssandraClust
+			if !utils.CompareAnnotations(actualDc, desiredDc, api.ResourceHashAnnotation) {
+				logger.Info("Updating datacenter")
+
+				if actualDc.Spec.SuperuserSecretName != desiredDc.Spec.SuperuserSecretName {
+					// If actualDc is created with SuperuserSecretName, it can't be changed anymore. We should reject all changes coming from K8ssandraCluster
+					desiredDc.Spec.SuperuserSecretName = actualDc.Spec.SuperuserSecretName
+					err = fmt.Errorf("tried to update superuserSecretName in K8ssandraCluster")
+					logger.Error(err, "SuperuserSecretName is immutable, reverting to existing value in CassandraDatacenter")
+				}
+
+				actualDc = actualDc.DeepCopy()
+				resourceVersion := actualDc.GetResourceVersion()
+				desiredDc.DeepCopyInto(actualDc)
+				actualDc.SetResourceVersion(resourceVersion)
+				if err = remoteClient.Update(ctx, actualDc); err != nil {
+					logger.Error(err, "Failed to update datacenter")
+					return result.Error(err), actualDcs
+				}
+			}
+
+			if !cassandra.DatacenterReady(actualDc) {
+				logger.Info("Waiting for datacenter to become ready")
+				return result.RequeueSoon(r.DefaultDelay), actualDcs
+			}
+
+			logger.Info("The datacenter is ready")
+
+			actualDcs = append(actualDcs, actualDc)
+
+			if recResult := r.updateReplicationOfSystemKeyspaces(ctx, kc, desiredDc, remoteClient, logger); recResult.Completed() {
+				return recResult, actualDcs
+			}
+
+		} else {
+			if errors.IsNotFound(err) {
+				// cassdc doesn't exist, we'll create it
+				if err = remoteClient.Create(ctx, desiredDc); err != nil {
+					logger.Error(err, "Failed to create datacenter")
+					return result.Error(err), actualDcs
+				}
+				return result.RequeueSoon(r.DefaultDelay), actualDcs
+			} else {
+				logger.Error(err, "Failed to get datacenter")
+				return result.Error(err), actualDcs
+			}
+		}
+	}
+
+	// If we reach this point all CassandraDatacenters are ready. We only set the
+	// CassandraInitialized condition if it is unset, i.e., only once. This allows us to
+	// distinguish whether we are deploying a CassandraDatacenter as part of a new cluster
+	// or as part of an existing cluster.
+	if kc.Status.GetConditionStatus(api.CassandraInitialized) == corev1.ConditionUnknown {
+		now := metav1.Now()
+		(&kc.Status).SetCondition(api.K8ssandraClusterCondition{
+			Type:               api.CassandraInitialized,
+			Status:             corev1.ConditionTrue,
+			LastTransitionTime: &now,
+		})
+	}
+
+	return result.Continue(), actualDcs
+}
+
+// checkSystemReplication checks for the SystemReplicationAnnotation on kc. If found, the
+// JSON value is unmarshalled and returned. If not found, the SystemReplication is computed
+// and is stored in the SystemReplicationAnnotation on kc. The value is JSON-encoded.
+// Lastly, kc is patched so that the changes are persisted,
+func (r *K8ssandraClusterReconciler) checkSystemReplication(ctx context.Context, kc *api.K8ssandraCluster, logger logr.Logger) (*cassandra.SystemReplication, error) {
```

Comment:
And again, I'd move this method and the next one to a `replication.go` file, wdyt?